spec/mailers/bulk_url_upload_results_mailer_spec.rb

```diff
@@ -93,6 +76,15 @@
       it 'shows the second URL validation failure' do
         expect(mail_body).to match(/#{second_error_message}\s+#{second_bad_url}/)
       end
+
+      it 'shows the urls with error "url already taken" after all the other errors' do
+        first_error_message_position = mail_body.index(first_error_message)
+        second_error_message_position = mail_body.index(second_error_message)
+        already_taken_error_message_position = mail_body.index(already_taken_error_message)
+
+        expect(already_taken_error_message_position).to be > first_error_message_position
+        expect(already_taken_error_message_position).to be > second_error_message_position
+      end
```

Comment:
@jmax-fearless , this is not a valid spec, because it passes without any of the changes in `spec/services/bulk_url_uploader_spec.rb`. The `before` block sets the errors in the correct order, so the spec is testing errors that are already in the expected order:
```ruby
      before do
        results.add_error(first_error_message, first_bad_url)
        results.add_error(second_error_message, second_bad_url)
        results.add_error(already_taken_error_message, duplicate_url)
      end
 ```

Please add a valid spec using TDD. As the results are being ordered in `BulkUrlUploader::Results`, you could add a unit test for that class, rather than testing this via the mailer.

Reply:
I'm not comfortable removing the mailer spec, but I agree that a unit test for this would be a fine thing. So I added one, and also re-arranged the order of `add_error` calls per your comment. Both test fail if you remove the changes to `#error_messages`